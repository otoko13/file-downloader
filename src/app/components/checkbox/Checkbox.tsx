"use client";

import { FunctionComponent } from "react";
import styles from "./checkbox.module.scss";
import classNames from "classnames";

export type CheckedState = "checked" | "partial" | "unchecked";

interface CheckboxProps {
  ariaLabel?: string;
  className?: string;
  checkedState: CheckedState;
  onClick?: () => void;
}

const Checkbox: FunctionComponent<CheckboxProps> = ({
  ariaLabel,
  checkedState,
  className,
  onClick,
}) => {
  const getCheckedAriaState = () => {
    switch (checkedState) {
      case "checked":
        return "true";
      case "partial":
        return "mixed";
      default:
        return "false";
    }
  };

  return (
    <button
      className={classNames(styles.checkbox, className, {
        [styles.active]: checkedState !== "unchecked",
      })}
      onClick={onClick}
      role="checkbox"
      aria-checked={getCheckedAriaState()}
      aria-label={ariaLabel ?? "Select"}
    >
      {checkedState === "checked" && (
        <i className={classNames("bi-check2", styles["check-icon"])} />
      )}
      {checkedState === "partial" && (
        <i className={classNames("bi-dash", styles["check-icon"])} />
      )}
    </button>
  );
};

export default Checkbox;
