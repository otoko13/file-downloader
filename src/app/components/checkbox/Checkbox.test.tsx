import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Checkbox from "./Checkbox";

describe("appearance", () => {
  it("should match the snapshot", () => {
    const { container } = render(<Checkbox checkedState="checked" />);
    expect(container).toMatchSnapshot();
  });
});

describe("checked state", () => {
  it("should show aria-checked as true when fully checked", () => {
    render(<Checkbox checkedState="checked" />);
    const checkbox = screen.getByRole("checkbox");
    expect(checkbox.getAttribute("aria-checked")).toBe("true");
  });

  it("should show aria-checked as mixes when parially checked", () => {
    render(<Checkbox checkedState="partial" />);
    const checkbox = screen.getByRole("checkbox");
    expect(checkbox.getAttribute("aria-checked")).toBe("mixed");
  });

  it("should show aria-checked as false when not checked", () => {
    render(<Checkbox checkedState="unchecked" />);
    const checkbox = screen.getByRole("checkbox");
    expect(checkbox.getAttribute("aria-checked")).toBe("false");
  });

  it("should show no icon when unchecked", () => {
    const { container } = render(<Checkbox checkedState="unchecked" />);
    const icons = container.querySelectorAll("i");
    expect(icons).toHaveLength(0);
  });

  it("should show a dash icon when partially checked", () => {
    const { container } = render(<Checkbox checkedState="partial" />);
    const icons = container.querySelectorAll("i");
    expect(icons).toHaveLength(1);
    expect(icons[0].className.includes("bi-dash"));
  });

  it("should show a check icon when fully checked", () => {
    const { container } = render(<Checkbox checkedState="checked" />);
    const icons = container.querySelectorAll("i");
    expect(icons).toHaveLength(1);
    expect(icons[0].className.includes("bi-check2"));
  });
});

describe("onClick", () => {
  it("should call onClick when clicked", async () => {
    const handler = jest.fn();
    render(<Checkbox checkedState="unchecked" onClick={handler} />);
    expect(handler).not.toHaveBeenCalled();
    await userEvent.click(screen.getByRole("checkbox"));
    expect(handler).toHaveBeenCalled();
  });
});
