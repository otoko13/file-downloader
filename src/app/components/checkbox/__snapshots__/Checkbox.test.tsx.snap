// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appearance should match the snapshot 1`] = `
<div>
  <button
    aria-checked="true"
    aria-label="Select"
    class="checkbox active"
    role="checkbox"
  >
    <i
      class="bi-check2 check-icon"
    />
  </button>
</div>
`;
